WebGL getUniformLocation, getAttribLocation calls are slow

kkinnunen-apple opened the pull request:
#### 6f718906cc3a3c3d943ff6997bbb920ec878025c
<pre>
WebGL getUniformLocation, getAttribLocation calls are slow
<a href="https://bugs.webkit.org/show_bug.cgi?id=300784">https://bugs.webkit.org/show_bug.cgi?id=300784</a>
<a href="https://rdar.apple.com/150998129">rdar://150998129</a>

Reviewed by Dan Glastonbury.

Calling getUniformLocation, getAttribLocation would need a synchronous
call to GPU process with very little payload work. Instead of repeated
sync calls, call activeUniforms(), activeAttribs() to get the full
uniform and attrib state. This way location getters as well as few other
calls can fetch the data only once.

The data invalidation was already present, invalidate on re-link.

The complexity in uniform location queries are the arrays, which
must be queriable by their base name as well as the full name with [0]
index. The array base names are added to uniform location map and
uniform index map as special case.
Attributes (100 es), vertex shader inputs (300 es) cannot be declared
as arrays, so they do not need the special casing.

Fix CString::toStdString() since it was incorrect. This patch adds
CString(const std::string&amp;) and this lets us add the correct
CString(const std::string&amp;) instead of replicating the toStdString bug
for consistency.

Speeds up
PerformanceTests/Canvas/WebGL-getAttribLocation-getUniformLocation.html
from ~1500 iterations to ~60000.
Speeds up webgl layout test runs.

Tests: Tools/TestWebKitAPI/Tests/WTF/CString.cpp
       Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp
* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(StringRange::StringRange):
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::currentExecutablePath):
* Source/WTF/wtf/text/CString.h:
(WTF::CString::CString):
(WTF::CString::toStdString const):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::getTransformFeedbackVarying):
(WebCore::WebGL2RenderingContext::getUniformIndices):
(WebCore::WebGL2RenderingContext::getActiveUniforms):
(WebCore::WebGL2RenderingContext::getActiveUniformBlockName):
* Source/WebCore/html/canvas/WebGLProgram.cpp:
(WebCore::WebGLProgram::linkStatus):
(WebCore::WebGLProgram::activeAttribs):
(WebCore::WebGLProgram::attribLocations):
(WebCore::WebGLProgram::activeUniforms):
(WebCore::WebGLProgram::uniformLocations):
(WebCore::WebGLProgram::uniformIndices):
(WebCore::WebGLProgram::requiredTransformFeedbackBufferCount):
(WebCore::WebGLProgram::increaseLinkCount):
(WebCore::WebGLProgram::getLinkStatus): Deleted.
(WebCore::WebGLProgram::cacheInfoIfNeeded): Deleted.
* Source/WebCore/html/canvas/WebGLProgram.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::getActiveAttrib):
(WebCore::WebGLRenderingContextBase::getActiveUniform):
(WebCore::WebGLRenderingContextBase::getAttribLocation):
(WebCore::WebGLRenderingContextBase::getProgramParameter):
(WebCore::WebGLRenderingContextBase::getUniform):
(WebCore::WebGLRenderingContextBase::getUniformLocation):
(WebCore::WebGLRenderingContextBase::useProgram):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsContextGLActiveInfo.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::activeAttribs):
(WebCore::GraphicsContextGLANGLE::activeUniforms):
(WebCore::GraphicsContextGLANGLE::getTransformFeedbackVarying):
(WebCore::GraphicsContextGLANGLE::getActiveUniforms): Deleted.
(WebCore::GraphicsContextGLANGLE::getActiveAttrib): Deleted.
(WebCore::GraphicsContextGLANGLE::getActiveUniform): Deleted.
(WebCore::GraphicsContextGLANGLE::getAttribLocation): Deleted.
(WebCore::GraphicsContextGLANGLE::getUniformLocation): Deleted.
(WebCore::GraphicsContextGLANGLE::getUniformIndices): Deleted.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGL::activeAttribs):
(WebKit::RemoteGraphicsContextGL::activeUniforms):
(WebKit::RemoteGraphicsContextGL::getTransformFeedbackVarying):
(WebKit::RemoteGraphicsContextGL::getActiveAttrib): Deleted.
(WebKit::RemoteGraphicsContextGL::getActiveUniform): Deleted.
(WebKit::RemoteGraphicsContextGL::getAttribLocation): Deleted.
(WebKit::RemoteGraphicsContextGL::getUniformLocation): Deleted.
(WebKit::RemoteGraphicsContextGL::getUniformIndices): Deleted.
(WebKit::RemoteGraphicsContextGL::getActiveUniforms): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::activeAttribs):
(WebKit::RemoteGraphicsContextGLProxy::activeUniforms):
(WebKit::RemoteGraphicsContextGLProxy::getTransformFeedbackVarying):
(WebKit::RemoteGraphicsContextGLProxy::getActiveAttrib): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getActiveUniform): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getAttribLocation): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getUniformLocation): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getUniformIndices): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getActiveUniforms): Deleted.
* Tools/Scripts/generate-gpup-webgl:
(main):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringNullStringConstructor)):
(TEST(WTF, CStringEmptyRegularConstructor)):
(TEST(WTF, CStringZeroTerminated)):
(TEST(WTF, CStringStdStringInterop)):
* Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp:
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteUint)):
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteNegativeInteger)):

Canonical link: <a href="https://commits.webkit.org/301863@main">https://commits.webkit.org/301863@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ca58ff64b86f0d3de7c3a7fde9e126b43165ca4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127314 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134378 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3afb7954-1436-4535-b622-f36f19b14456) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129186 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47572 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55484 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d5ded704-42d9-406b-aa74-39ff51046595) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38062 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114024 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77390 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c7b51518-ecf4-4308-a31a-c9819c6ac672) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/126665 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77760 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119351 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107924 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136861 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125777 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53973 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41579 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105413 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54484 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110374 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105095 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26787 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50611 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51541 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53910 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59997 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158815 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39713 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54903 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -1316,38 +1307,81 @@ void GraphicsContextGLANGLE::generateMipmap(GCGLenum target)
+    GLsizei activeAttribs = 0;
+    GL_GetProgramiv(program, GL_ACTIVE_ATTRIBUTES, &activeAttribs);
+    std::string name;
+    Vector<GCGLAttribActiveInfo> result;
```
Should you request the capacity of `result` to be `activeAttribs`?

djg commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -1316,38 +1307,81 @@ void GraphicsContextGLANGLE::generateMipmap(GCGLenum target)
+    GLsizei activeUniforms = 0;
+    GL_GetProgramiv(program, GL_ACTIVE_UNIFORMS, &activeUniforms);
+    std::string name;
+    Vector<GCGLUniformActiveInfo> result;
```
ditto

djg commented on Source/WebCore/html/canvas/WebGLProgram.h:
```diff
@@ -92,21 +95,32 @@ class WebGLProgram final : public WebGLObject, public ContextDestructionObserver
+        bool hasAttribLocations { false };
+        bool hasActiveUniforms { false };
+        bool hasUniformLocations { false };
+        bool hasUniformIndices { false };
```
Instead of keeping separate state for "having a valid value", Would these be better as `std::optional`?

djg commented on Source/WebCore/html/canvas/WebGLProgram.cpp:
```diff
@@ -111,16 +111,108 @@ void WebGLProgram::deleteObjectImpl(const AbstractLocker& locker, GraphicsContex
+        for (auto& activeUniform : activeUniforms()) {
+            auto name = String::fromUTF8(activeUniform.name.data());
+            locations.add(name, activeUniform.locations[0]);
+            if (name.endsWith("[0]"_s)) {
```
This `[0]` suffix is removed in GraphicsContextGLANGLE::activeUniforms(). Should it be done only in one place; If we have to support non-Cocoa contexts, does it make sense to do it here?

kkinnunen-apple replied:
It is not removed in GraphicsContextGLANGLE. There it's removed to form the names to query the uniform locations of the subsequent array indices.

Here the condition is used to identify an array. E.g. "This active uniform is an array, so make it possible to query the array location with the base name in addition to fully qualified array first element name"

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
WebGL getUniformLocation, getAttribLocation calls are slow
https://bugs.webkit.org/show_bug.cgi?id=300784
rdar://150998129

Reviewed by Dan Glastonbury.

Calling getUniformLocation, getAttribLocation would need a synchronous
call to GPU process with very little payload work. Instead of repeated
sync calls, call activeUniforms(), activeAttribs() to get the full
uniform and attrib state. This way location getters as well as few other
calls can fetch the data only once.

The data invalidation was already present, invalidate on re-link.

The complexity in uniform location queries are the arrays, which
must be queriable by their base name as well as the full name with [0]
index. The array base names are added to uniform location map and
uniform index map as special case.
Attributes (100 es), vertex shader inputs (300 es) cannot be declared
as arrays, so they do not need the special casing.

Fix CString::toStdString() since it was incorrect. This patch adds
CString(const std::string&) and this lets us add the correct
CString(const std::string&) instead of replicating the toStdString bug
for consistency.

Speeds up
PerformanceTests/Canvas/WebGL-getAttribLocation-getUniformLocation.html
from ~1500 iterations to ~60000.
Speeds up webgl layout test runs.

Tests: Tools/TestWebKitAPI/Tests/WTF/CString.cpp
       Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp
* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(StringRange::StringRange):
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::currentExecutablePath):
* Source/WTF/wtf/text/CString.h:
(WTF::CString::CString):
(WTF::CString::toStdString const):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::getTransformFeedbackVarying):
(WebCore::WebGL2RenderingContext::getUniformIndices):
(WebCore::WebGL2RenderingContext::getActiveUniforms):
(WebCore::WebGL2RenderingContext::getActiveUniformBlockName):
* Source/WebCore/html/canvas/WebGLProgram.cpp:
(WebCore::WebGLProgram::linkStatus):
(WebCore::WebGLProgram::activeAttribs):
(WebCore::WebGLProgram::attribLocations):
(WebCore::WebGLProgram::activeUniforms):
(WebCore::WebGLProgram::uniformLocations):
(WebCore::WebGLProgram::uniformIndices):
(WebCore::WebGLProgram::requiredTransformFeedbackBufferCount):
(WebCore::WebGLProgram::increaseLinkCount):
(WebCore::WebGLProgram::getLinkStatus): Deleted.
(WebCore::WebGLProgram::cacheInfoIfNeeded): Deleted.
* Source/WebCore/html/canvas/WebGLProgram.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::getActiveAttrib):
(WebCore::WebGLRenderingContextBase::getActiveUniform):
(WebCore::WebGLRenderingContextBase::getAttribLocation):
(WebCore::WebGLRenderingContextBase::getProgramParameter):
(WebCore::WebGLRenderingContextBase::getUniform):
(WebCore::WebGLRenderingContextBase::getUniformLocation):
(WebCore::WebGLRenderingContextBase::useProgram):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsContextGLActiveInfo.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::activeAttribs):
(WebCore::GraphicsContextGLANGLE::activeUniforms):
(WebCore::GraphicsContextGLANGLE::getTransformFeedbackVarying):
(WebCore::GraphicsContextGLANGLE::getActiveUniforms): Deleted.
(WebCore::GraphicsContextGLANGLE::getActiveAttrib): Deleted.
(WebCore::GraphicsContextGLANGLE::getActiveUniform): Deleted.
(WebCore::GraphicsContextGLANGLE::getAttribLocation): Deleted.
(WebCore::GraphicsContextGLANGLE::getUniformLocation): Deleted.
(WebCore::GraphicsContextGLANGLE::getUniformIndices): Deleted.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGL::activeAttribs):
(WebKit::RemoteGraphicsContextGL::activeUniforms):
(WebKit::RemoteGraphicsContextGL::getTransformFeedbackVarying):
(WebKit::RemoteGraphicsContextGL::getActiveAttrib): Deleted.
(WebKit::RemoteGraphicsContextGL::getActiveUniform): Deleted.
(WebKit::RemoteGraphicsContextGL::getAttribLocation): Deleted.
(WebKit::RemoteGraphicsContextGL::getUniformLocation): Deleted.
(WebKit::RemoteGraphicsContextGL::getUniformIndices): Deleted.
(WebKit::RemoteGraphicsContextGL::getActiveUniforms): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::activeAttribs):
(WebKit::RemoteGraphicsContextGLProxy::activeUniforms):
(WebKit::RemoteGraphicsContextGLProxy::getTransformFeedbackVarying):
(WebKit::RemoteGraphicsContextGLProxy::getActiveAttrib): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getActiveUniform): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getAttribLocation): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getUniformLocation): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getUniformIndices): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::getActiveUniforms): Deleted.
* Tools/Scripts/generate-gpup-webgl:
(main):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringNullStringConstructor)):
(TEST(WTF, CStringEmptyRegularConstructor)):
(TEST(WTF, CStringZeroTerminated)):
(TEST(WTF, CStringStdStringInterop)):
* Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp:
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteUint)):
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteNegativeInteger)):

Canonical link: https://commits.webkit.org/301863@main
